list tracker refactor

heronhaye opened the pull request:
Overall goal: Switch following lists to be sigchain-verified and not servertrust (with the caveat that server can arbitrarily remove items from the list), and make RPCs simpler.


. Refactors, renames, and deletes several structs and variables to be more clear
. Changes ListTracking to get data from unstubbed UPAK instead of FullSelfer, in order to take advantage of UPAK caching. Accordingly, output less information in CLI `list-following`.
. Unifies ListTracking/ListTracking2, UserSummary/UserSummary2
. Removes ability to (easily) get unverified following information for any user
. Tries to maintain TrackerLoader behavior where cached data is immediately returned
. Moves trackerloader to service so it can use engine
. Sources state.config.following from sigchain now (with caveat that resets are not checked). state.config.followers is still not verified.
. Removes unneeded GUI manipulation of state.config.followers/following

heronhaye commented on go/engine/list_trackers.go:
```diff
@@ -0,0 +1,108 @@
+// Copyright 015 Keybase, Inc. All rights reserved. Use of
+// this source code is governed by the included BSD license.
+
+package engine
```
This file moved from list_trackers2.go with some changes.

heronhaye commented on go/service/tracker_loader.go:
```diff
@@ -1,17 +1,19 @@
-package libkb
+package service
```
Moved to Service so it can use Engines

heronhaye commented on go/identify3/adapter.go:
```diff
@@ -554,14 +554,6 @@ func (i *UIAdapter) DisplayTrackStatement(libkb.MetaContext, string) error {
 
 func (i *UIAdapter) DisplayUserCard(mctx libkb.MetaContext, card keybase1.UserCard) error {
-
-	// Do not take the server's word on this! Overwrite with what we got above.
```
Get rid of this field completely so we don't accidentally mess this up.

heronhaye commented on go/libkb/sync_trackers.go:
```diff
@@ -11,28 +11,39 @@ import (
+	return direction == FollowDirectionFollowing
+}
+
+type ServertrustTrackerSyncer struct {
```
Just some renaming in this file. No behavior changes.

heronhaye commented on go/libkb/user_card.go:
```diff
@@ -5,20 +5,14 @@ import (
 )
 
 type card struct {
-	Status        AppStatus `json:"status"`
```
I removed fields that are obtainable verified (and modified users)

heronhaye commented on go/libkb/upak_loader.go:
```diff
@@ -32,7 +32,6 @@ type UPAKLoader interface {
 	LookupUsernameUPAK(ctx context.Context, uid keybase1.UID) (NormalizedUsername, error)
 	LookupUID(ctx context.Context, un NormalizedUsername) (keybase1.UID, error)
 	LookupUsernameAndDevice(ctx context.Context, uid keybase1.UID, did keybase1.DeviceID) (username NormalizedUsername, deviceName string, deviceType string, err error)
-	ListFollowedUIDs(ctx context.Context, uid keybase1.UID) ([]keybase1.UID, error)
```
This didn't seem to be correct. Maybe it's because it's loading Unstubbed? I just removed it since it wasn't used anyway.

heronhaye commented on protocol/avdl/keybase1/user.avdl:
```diff
@@ -86,38 +112,12 @@ protocol user {
     */
   array<PublicKey> loadAllPublicKeysUnverified(int sessionID, UID uid);
 
-  record UserSummary2 {
```
I unified this with UserSummary, but kept it compatible with the UserSummary2 that is in our leveldbs. Removed isFollower/isFollowee fields.

heronhaye commented on shared/reducers/config.tsx:
```diff
@@ -33,22 +33,6 @@ export default Container.makeReducer<Actions, Types.State>(Constants.initialStat
       ).username
     }
   },
-  [Tracker2Gen.updatedDetails]: (draftState, action) => {
```
This comes in from Identify3 but we don't need it, following and followers should be kept up to date outside of the identify system.

heronhaye commented on shared/reducers/tracker2.tsx:
```diff
@@ -49,8 +49,6 @@ export default Container.makeReducer<Actions, Types.State>(initialState, {
     const d = getDetails(draftState, username)
     d.bio = action.payload.bio
     d.blocked = action.payload.blocked
-    d.followersCount = action.payload.followersCount
```
Will be set later when follows are loaded

heronhaye replied:
(meanwhile there's a spinner)

chrisnojima replied:
but we don't ever load these in the tracker right?

heronhaye replied:
Good point. Can we hide this in the tracker then?

chrisnojima replied:
this was the motivation for having the count and the list being 2 things. i'd ask cecile

heronhaye commented on shared/constants/types/profile.tsx:
```diff
@@ -2,16 +2,6 @@ import * as RPCTypes from './rpc-gen'
 
-export type FriendshipsTab = 'Followers' | 'Following'
-
-export type FriendshipUserInfo = {
```
Seemed unused

heronhaye commented on go/client/cmd_upak.go:
```diff
@@ -18,9 +18,9 @@ import (
 	libkb.Contextified
-	uid  keybase1.UID
-	kid  keybase1.KID
-	lite bool
```
lite isn't really used in the app, just killed it

heronhaye commented on go/engine/pgp_pull.go:
```diff
@@ -207,16 +193,47 @@ func (e *PGPPullEngine) runLoggedIn(m libkb.MetaContext) error {
 
+	pkLookup := make(map[keybase1.UID][]string)
+
+	err = m.G().GetFullSelfer().WithSelf(m.Ctx(), func(user *libkb.User) error {
```
Because the UPAK data from list following doesn't include it, need to do a FullSelfer load to get the relevant data now.

heronhaye commented on go/engine/user_summary.go:
```diff
@@ -1,223 +0,0 @@
-// Copyright 2015 Keybase, Inc. All rights reserved. Use of
```
Deleted all this stuff, didn't seem to be used

heronhaye commented on go/libkb/upak_loader.go:
```diff
@@ -377,10 +376,15 @@ func (u *CachedUPAKLoader) loadWithInfo(arg LoadUserArg, info *CachedUserLoadInf
 
-	lock := u.locktab.AcquireOnName(ctx, g, arg.uid.String())
+	var lock *NamedLock
+	if !arg.cachedOnly {
```
Don't need to singleflight if we're only loading cached data without calls to network.

heronhaye commented:
cc @patrickxb, I changed a lot of code

chrisnojima commented on shared/actions/tracker2.tsx:
```diff
@@ -279,7 +303,8 @@ function* tracker2Saga() {
   yield* Saga.chainAction(Tracker2Gen.ignore, ignore)
   yield* Saga.chainGenerator<Tracker2Gen.LoadPayload>(Tracker2Gen.load, load)
-  yield* Saga.chainAction(Tracker2Gen.load, loadFollow)
+  yield* Saga.chainAction(Tracker2Gen.load, loadFollowers)
```
is there some motivation for splitting it into 2 side effects now. before it had one side effect and made 2 calls and joined them. This could still do the same

heronhaye replied:
With my change, if you're loading someone with a large sigchain like chris for the first time, loading the following list takes a lot longer than the followers list. I split it up so the followers tab can populate first. I can change it if there's a better way to do the same thing

heronhaye commented:
@chrisnojima 
Talked to her and she said that we want it. So I brought back the unverified counts from the usercard and use that for tracker (but it's overridden in profile page).
I also fixed an issue where full names were not being threaded through. Tsc caught it.

Finally I fixed a pgp pull test bug.

AMarcedone commented on go/engine/list_tracking.go:
```diff
@@ -59,149 +54,123 @@ func (e *ListTrackingEngine) RequiredUIs() []libkb.UIKind { return []libkb.UIKin
-	for _, trackedKey := range trackedKeys {
-		res = append(res, keybase1.PublicKey{PGPFingerprint: trackedKey.Fingerprint.String()})
+	sort.Slice(filteredTracks, func(i, j int) bool {
+		return strings.Compare(filteredTracks[i].Username, filteredTracks[j].Username) < 0
```
You probably don't want `strings.Compare` https://stackoverflow.com/questions/52830314/go-compare-strings/52830378
Simple `<` is better

heronhaye replied:
TIL, thanks. That's odd.

heronhaye commented on go/engine/list_trackers.go:
```diff
@@ -0,0 +1,113 @@
+}
+
+// GetPrereqs returns the engine prereqs (none).
+func (e *ListTrackersUnverifiedEngine) Prereqs() Prereqs {
```
Neither am I to be honest, but I think it's just a bad comment. This basically means "If the engine was not provided a specific assertion or uid to look up, use the logged in uid. In that case we need a session).

heronhaye commented on go/engine/list_tracking_test.go:
```diff
@@ -158,29 +127,140 @@ func TestListTrackingLocal(t *testing.T) {
+
+	ResetAccount(btc, bob)
+
+	// Due to a server bug, it does the follow version doesn't update
```
Mentioned ticket # in comment

heronhaye commented on go/libkb/user_card.go:
```diff
@@ -63,21 +61,19 @@ func UserCard(m MetaContext, uid keybase1.UID, useSession bool) (ret *keybase1.U
+		Location:               card.Profile.Location,
+		Bio:                    card.Profile.Bio,
+		Website:                card.Profile.Website,
+		Twitter:                card.Profile.Twitter,
```
Ticketed

heronhaye commented on go/service/config.go:
```diff
@@ -339,9 +339,12 @@ func (h ConfigHandler) GetBootstrapStatus(ctx context.Context, sessionID int) (k
 }
 
-func (h ConfigHandler) RequestFollowerInfo(ctx context.Context, uid keybase1.UID) error {
+func (h ConfigHandler) RequestFollowingAndUnverifiedFollowers(ctx context.Context, sessionID int) error {
```
It's used for threading together several different RPCs. For example, I think crypto in the gui might use it so the GUI knows what operation things like encryption progress notifications are for (this is just my understanding, I've never used them myself, it's usually unused).

heronhaye commented on go/service/user.go:
```diff
@@ -45,40 +45,50 @@ func NewUserHandler(xp rpc.Transporter, g *libkb.GlobalContext, chatG *globals.C
-}
-
-func (h *UserHandler) ListTracking(ctx context.Context, arg keybase1.ListTrackingArg) (res []keybase1.UserSummary, err error) {
+func (h *UserHandler) ListTracking(ctx context.Context, arg keybase1.ListTrackingArg) (ret keybase1.UserSummarySet, err error) {
```
Will do in a subsequent PR.
